proposals/simd-XXXX-vm-consume-budget-for-percise-failure.md

```diff
@@ -0,0 +1,76 @@
+
+### Clarified Protocol Behavior:
+
+Instead of mandating implementation-specific work to handle exceptions, we propose the following clarification in the protocol:
+
+- For successful execution of a Basic Block (i.e., the block exits at the last BPF instruction), the deterministic CU cost of the block will be charged to the transaction’s CU meter. This ensures that CU consumption for successful transactions is accurately accounted for.
+- In the event of an exception during Basic Block execution, where the block does not exit normally, the requested CUs for the transaction will be charged to the CU meter. This allows for a simple and efficient fallback mechanism that avoids the need for tracking the exact number of executed instructions up to the point of failure.
+
+By adopting this approach, the protocol avoids the overhead of requiring precise instruction-level CU tracking for transactions that fail. Instead, the requested CU limit of the transaction will be used, simplifying the handling of failed transactions while still maintaining consensus.
+
+### Conclusion:
+
+This proposal enhances performance and simplifies CU tracking by formalizing the use of Basic Blocks for efficient execution. It eliminates the need for costly, implementation-specific work to track CU consumption during execution failures, providing a clear and consistent approach to handling exceptions. This change allows clients to maintain consensus without sacrificing performance, ensuring that the protocol remains both efficient and robust.
+
+## Alternatives Considered
+
+None
+
+## New Terminology
+
+- [Basic Block](https://en.wikipedia.org/wiki/Basic_block):i In the context of JIT execution and BPF processing, a Basic Block is a sequence of BPF instructions that forms a single, linear flow of control with no loops or conditional branches except for the entry and exit points. It represents a segment of code where execution starts at the first instruction and proceeds sequentially through to the last instruction without deviation. The Basic Block is characterized by its predictable execution path, allowing for efficient budget checks and optimizations, as its Compute Unit (CU) cost can be determined before execution and verified at the end of the block.
+
+## Detailed Design
+
+At banking stage [here](https://github.com/anza-xyz/agave/blob/master/core/src/banking_stage/committer.rs#L99) and replay stage [here](https://github.com/anza-xyz/agave/blob/master/ledger/src/blockstore_processor.rs#L239) where Transaction's executed_units is checked, implement new logic:
+```
+let execution_cu = match transaction.execution_results {
+    Ok(_) || Err(TransactionError::CustomError(_)) => committed_tx.executed_cu,
+    _ => transaction.requested_cu,
+};
```

Comment:
Not too familiar with InstructionError, wondering if correct to use Requested_cu for any error but `TransactionError::CustomError(_)`

Reply:
This is the most important part to get right. In Agave's VM, I think it's
```
EbpfError::DivideByZero
EbpfError::DivideOverflow
EbpfError::CallOutsideTextSegment
EbpfError::InvalidInstruction
EbpfError::InvalidVirtualAddress
```
I'm not sure exactly what these translate to as `TransactionError`s though.
In firedancer, it's maybe something like
```
#define FD_VM_ERR_SIGSPLIT    ( -9) /* split multiword instruction (e.g. jump into the middle of a multiword instruction) */
#define FD_VM_ERR_SIGILL      (-12) /* illegal instruction (e.g. opcode is not valid) */
#define FD_VM_ERR_SIGSEGV     (-13) /* illegal memory address (e.g. read/write to an address not backed by any memory) */
#define FD_VM_ERR_SIGBUS      (-14) /* misaligned memory address (e.g. read/write to an address with inappropriate alignment) */
#define FD_VM_ERR_SIGRDONLY   (-15) /* illegal write (e.g. write to a read only address) */
#define FD_VM_ERR_SIGFPE      (-18) /* divide by zero */
```

Reply:
indeed, what about `AccessViolation` in agave's VM, should it also be considered as "irregular failure"?

I failed to find where EbpfError converted to InstructionError, there are some casting at `bpf_loader`, is it the right place to look at @Lichtso ?

Reply:
It happens here: https://github.com/anza-xyz/agave/blob/7741b250a6e76afc9e7385ceb64c341f4bc21622/programs/bpf_loader/src/lib.rs#L1473